### PR TITLE
Implement masked batch normalization

### PR DESCRIPTION
**Pull request by NeoLegends**
We commonly use the regular torch `nn.BatchNorm1d`, which does not mask non-sequence positions when computing the batch statistics. Masking them away is more correct.

Based on https://gist.github.com/amiasato/902fc14afa37a7537386f7b0c5537741.

### CODE REVIEW

File: i6_models/parts/masked_norm.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+        # we use the mask to calculate the mean
+        mask = mask / n
+        mask = mask.unsqueeze(-2)
+
+        reduce_dims = list(range(inp.ndim - 2)) + [inp.ndim - 1]
+        if not self.track_running_stats:
+            mean = (mask * inp).sum(reduce_dims)
+            var = ((mask * inp**2).sum(reduce_dims) - mean**2) * n / (n - 1)
+        elif self.training and n > 1:
```
Comment (kuacakuaca) [inline i6_models/parts/masked_norm.py:95]: you don't need n > 1 check anymore.

File: i6_models/parts/masked_norm.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+import torch
+from torch import nn, Tensor
+
+
+def _lengths_to_mask(lengths: Tensor, max_len=None, dtype=None) -> Tensor:
+    """
+    Converts a "lengths" tensor to its binary mask representation.
+    """
+    assert len(lengths.shape) == 1, "Length shape should be 1 dimensional."
```
Comment (kuacakuaca) [inline i6_models/parts/masked_norm.py:13]: does this have to be one-dimensional? if so, then the dimensionality of `inp` in the forward function is also fixed.

**Review by kuacakuaca — CHANGES_REQUESTED** [review]

File: i6_models/parts/masked_norm.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+        mask = mask.unsqueeze(-2)
+
+        reduce_dims = list(range(inp.ndim - 2)) + [inp.ndim - 1]
+        if not self.track_running_stats:
+            mean = (mask * inp).sum(reduce_dims)
+            var = ((mask * inp**2).sum(reduce_dims) - mean**2) * n / (n - 1)
+        elif self.training:
+            mean = (mask * inp).sum(reduce_dims)
+            # Var(X) = E[X^2] - E[X]^2
+            var = (mask * inp**2).sum(reduce_dims) - mean**2
```
Comment (michelwi) [inline i6_models/parts/masked_norm.py:109]: Why is there a distinction of having/not-having the factor `n / (n - 1)` in training vs (not training or not track_running_stats)? especially if we add the factor again to the running_var that we are tracking.

File: i6_models/parts/masked_norm.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+
+            with torch.no_grad():
+                self.running_mean = (
+                    exponential_average_factor * mean + (1 - exponential_average_factor) * self.running_mean
+                )
+                # torch updates running statistics with unbiased var
+                self.running_var = (
+                    exponential_average_factor * var * n / (n - 1) + (1 - exponential_average_factor) * self.running_var
```
Comment (michelwi) [inline i6_models/parts/masked_norm.py:117]: is exponentially averaging the variances the correct thing to do? The result will be different from the variance of the whole data.. e.g.
```
# prepare 10 batches of random data, each shifted by a different amount
datas = [torch.tensor(np.random.random(data.shape) + i) for i in range(10)]
norm = MaskedBatchNorm1dV1(num_features=data.shape[1], momentum=None, affine=False).train()
_ = [norm(d, data_lens) for d in datas]

norm.running_var, norm.running_mean  # 0.0856, 4.9849
torch.var_mean(torch.cat(datas))     # 8.3816, 4.9849
```
In practice I doubt the inter-batch variance will be much larger than the intra-batch variance.. so maybe we do not need to be concerned?

File: i6_models/parts/masked_norm.py [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+        super().__init__(
+            num_features,
+            eps=eps,
+            momentum=momentum,
+            affine=affine,
+            track_running_stats=track_running_stats,
+            device=device,
+            dtype=dtype,
+        )
+
```
[the remark this answers is not in the document]
Comment (NeoLegends) [reply i6_models/parts/masked_norm.py:61]: IIRC torch modules require a constructor to be defined.

File: i6_models/parts/masked_norm.py (hunk 3 above)
Comment (NeoLegends) [reply i6_models/parts/masked_norm.py:109]: Citing torch docs: 

> At train time in the forward pass, the variance is calculated via the biased estimator, equivalent to torch.var(input, unbiased=False). However, the value stored in the moving average of the variance is calculated via the unbiased estimator, equivalent to torch.var(input, unbiased=True).

File: i6_models/parts/masked_norm.py [hunk 6]
```diff
@@ -0,0 +1,127 @@
+__all__ = ["MaskedBatchNorm1dV1"]
+
+from typing import Optional
+
+import torch
+from torch import nn, Tensor
+
+
+def _lengths_to_mask(lengths: Tensor, max_len: Optional[int] = None, dtype=None) -> Tensor:
```
[the remark this answers is not in the document]
Comment (NeoLegends) [reply i6_models/parts/masked_norm.py:9]: I'd leave this to a follow-up.

**Review by kuacakuaca — APPROVED** [review]
